scout/run_batch.py

```diff
@@ -0,0 +1,131 @@
+from __future__ import annotations
+from pathlib import Path
+from scout.config import Config, LogConfig, FilePaths as fp  # noqa: F401
+from scout.ecm_prep_args import ecm_args
+from scout import ecm_prep, run
+from argparse import ArgumentParser
+import logging
+logger = logging.getLogger(__name__)
+
+
+class BatchRun():
+    def __init__(self, yml_dir):
+        self.yml_dir = yml_dir.resolve()
+
+    def get_ecm_files(self, ymls: list) -> list:  # noqa: F821
+        """Retrieve all ECMs from 1 or more config file and concatenate into a single list
+
+        Args:
+            ymls (list): filepaths of yml configuration files
+
+        Returns:
+            list: list of all ECMs specified in the yml files
+        """
+
+        ecm_files = set()
+        for config in ymls:
+            config_pth = str(config.resolve())
+            ecm_prep_opts = ecm_args(["-y", config_pth])
+            ecm_files.update(set(ecm_prep_opts.ecm_files))
+
+        return list(ecm_files)
+
+    def get_run_opts(self, config_pth: Path) -> argparse.NameSpace:  # noqa: F821
+        """Parse arguments for run.py for a given configuration file. If a results_directory
+            argument is not specified in the yml, then the name of the yml will be set as the
+            results directory.
+
+        Args:
+            config_pth (Path): path to the yml configuration file
+
+        Returns:
+            argparse.NameSpace: arguments to be used in run.py
+        """
+
+        results_dir_yml = Config.load_config(config_pth)["run"].get("results_directory", None)
+        if results_dir_yml:
+            run_opts = run.parse_args(["-y", str(config_pth.resolve())])
+        else:
+            custom_results_dir = Path(__file__).resolve().parents[1] / "results" / config_pth.stem
+            run_opts = run.parse_args(["-y", str(config_pth.resolve()),
+                                      "--results_directory", str(custom_results_dir)])
+
+        return run_opts
+
+    def group_common_configs(self, yml_dir: Path):
+        """Groups together configuration files with similar ecm_prep arguments. Those with
+            identical ecm_prep arguments excluding `ecm_directory`, `ecm_files`, and
+            `ecm_files_regex` are put in the same group to optimize running ecm_prep.py.
+            Note: those with different `ecm_packages` values are not grouped together, as this
+            influences which individual ECMs are run, and may omit preparation of desired
+            measures.
+
+        Args:
+            yml_dir (Path): directory containing all configurations with which to run.
+
+        Returns:
+            list: groups of similar yamls that can be run with a single ecm_prep.py call
+        """
+
+        exclude_keys = ["ecm_directory", "ecm_files", "ecm_files_regex"]
```

Comment:
Given the latest assumptions built into https://github.com/trynthink/scout/pull/372, I think these three still make sense as excluded keys. We do _not_ want to group those with differing ecm_packages, because that affects which ECMs are run individually. Conversely, any differences in ecm_files (or ecm_regex/ecm_dir) will not affect packages.